src/gravel/controllers/services.py

```diff
@@ -152,6 +185,58 @@ def get(self, name: str) -> ServiceModel:
             raise UnknownServiceError(name)
         return self._services[name]
 
+    @property
+    def constraints(self) -> ConstraintsModel:
+        """
+        Calculate constraints for service deployment. These are generic and not
+        tied to any particular service's requirements.
+        """
+        devices_ctrl: Devices = get_devices()
+        devs_per_host: Dict[str, DeviceHostModel] = \
+            devices_ctrl.devices_per_host
+
+        logger.debug(f"get constraints, hosts = {devs_per_host}")
+
+        hosts: List[str] = \
+            [h for h, d in devs_per_host.items() if len(d.devices) > 0]
+        num_hosts: int = len(hosts)
+
+        availability: AvailabilityConstraints = \
+            AvailabilityConstraints(hosts=num_hosts)
+
+        max_devs: int = 0
+        min_devs: int = -1
+        for h in hosts:
```

Comment:
(nit) instead of constructing hosts on line 200, the AvailabilityConstraints and appropriate min/max parts could be done in the same loop.